core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -338,27 +340,27 @@ synchronized final void completeExceptionally(Throwable throwable, int topologyI
 
    private class SingleKeyCollector extends Collector<Object> {
       @GuardedBy("this")
-      private final Collection<Address> owners;
+      private final HashSet<Address> owners;
       @GuardedBy("this")
       private final Address primaryOwner;
       @GuardedBy("this")
       private Object returnValue;
 
-      private SingleKeyCollector(CommandInvocationId id, Collection<Address> owners, int topologyId) {
-         super(id, topologyId);
+      private SingleKeyCollector(Long id, Collection<Address> owners, int topologyId, long timeoutNanos) {
+         super(id, topologyId, timeoutNanos);
          this.primaryOwner = owners.iterator().next();
          this.owners = new HashSet<>(owners); //removal is fast
       }
 
-      private SingleKeyCollector(CommandInvocationId id, Object returnValue, Collection<Address> owners,
-            int topologyId) {
-         super(id, topologyId);
+      private SingleKeyCollector(Long id, Object returnValue, Collection<Address> owners,
+            int topologyId, long timeoutNanos) {
+         super(id, topologyId, timeoutNanos);
          this.returnValue = returnValue;
          this.primaryOwner = owners.iterator().next();
-         Collection<Address> tmpOwners = new HashSet<>(owners);
+         HashSet<Address> tmpOwners = new HashSet<>(owners);
```

Comment:
I think it would be cheaper to receive the list of backups as a parameter, especially after https://github.com/infinispan/infinispan/pull/4804 goes in.